### ISSUE TITLE
libnd4j: need trace backprop op

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#6078]
https://www.tensorflow.org/api_docs/python/tf/trace

2d case is easy: it's just eye(input shape).mul(gradient) (where 'gradient' is a scalar).

For higher dimensional case: say rank 4 input, shape [a,b,c,d]
output (forward) is rank 4, shape [a,b], hence gradientAtOutput is shape [a,b]
GradAtInput is just [a,b,c,d] with the diagonal of [a[i],b[j],:,:] all set to single value gradAtOutput[a[i],b[j]] with off-diagonal values being 0.

For the 5d case, it'd be [a,b,c,d,e] in, [a,b,c] gradOut, then set the diagonal of [a[i],b[j],c[k],:,:] to gradAtOutput[a[i],b[j],c[k]]

**Comment by AlexDBlack** [comment]
Alternatively: if I had this op, I could do it easily in Java only - https://github.com/deeplearning4j/deeplearning4j/issues/6079

**Comment by AlexDBlack** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/6080
https://github.com/deeplearning4j/deeplearning4j/issues/6079
